File: TemplateMethodPattern/src/Beverage/Beverage.java
package beverage;

public abstract class Beverage {
	public final void prepareRecipe() {
		boilWater();
		brew();
		pourInCup();
		add();
	}

	abstract protected void add();

	abstract protected void brew();
	
	public void boilWater() {
		System.out.println("Boiling Water");
	}
	
	public void pourInCup() {
		System.out.println("Pouring into cup");
	}

}

File: NVIPattern/NVIPattern/Beverage.cpp
#include "Beverage.h"

using namespace std;

Beverage::Beverage(void)
{
}

void Beverage::prepareRecipe() {
	boilWater();
	brew();
	pourInCup();
	add();
}

Beverage::~Beverage(void)
{
}

void Beverage::boilWater()
{
	cout << "Boiling Water\n";
}

void Beverage::pourInCup()
{
	cout << "Pouring into cup\n";
}

void Beverage::add()
{
	doAdd();
}

void Beverage::brew()
{
	if(checkResourceQuality()) {
		doBrew();
	}
}
